### PR TITLE
Add IConstraintFactory

### PR DESCRIPTION
**Pull request by jbagga**
Add ConstraintResolver infrastructure to Dispatcher to enable porting `TreeMatcher`

Part of #472

### CODE REVIEW

File: src/Microsoft.AspNetCore.Dispatcher/Constraints/CompositeRouteConstraint.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    /// <summary>
+    /// Constrains a route by several child constraints.
+    /// </summary>
+    public class CompositeRouteConstraint : IDispatcherValueConstraint
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Constraints/CompositeRouteConstraint.cs]: Make the names match

File: src/Microsoft.AspNetCore.Dispatcher/Constraints/CompositeRouteConstraint.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    /// <summary>
+    /// Constrains a route by several child constraints.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Constraints/CompositeRouteConstraint.cs]: what's a route??

File: src/Microsoft.AspNetCore.Dispatcher/Constraints/DefaultConstraintResolver.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+            if (!_constraintMap.TryGetValue(constraintKey, out var constraintType))
+            {
+                // Cannot resolve the constraint key
+                return null;
+            }
+
+            if (!typeof(IDispatcherValueConstraint).GetTypeInfo().IsAssignableFrom(constraintType.GetTypeInfo()))
+            {
+                throw new RouteCreationException(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Constraints/DefaultConstraintResolver.cs]: What's a route???

File: src/Microsoft.AspNetCore.Dispatcher/Constraints/DispatcherValueConstraintMatcher.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    public static class DispatcherValueConstraintMatcher
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Constraints/DispatcherValueConstraintMatcher.cs]: Is this class used anywhere in the new code? I'd like to just not port this.

File: src/Microsoft.AspNetCore.Dispatcher/Constraints/IConstraintResolver.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    /// <summary>
+    /// Defines an abstraction for resolving constraints as instances of <see cref="IDispatcherValueConstraint"/>.
+    /// </summary>
+    public interface IConstraintResolver
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Constraints/IConstraintResolver.cs]: call this `IConstraintFactory`

File: src/Microsoft.AspNetCore.Dispatcher/RouteCreationException.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Dispatcher
+{
+    /// <summary>
+    /// The exception that is thrown for invalid routes or constraints.
+    /// </summary>
+    public class RouteCreationException : Exception
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/RouteCreationException.cs]: We don't need this type

Comment (jbagga) [reply src/Microsoft.AspNetCore.Dispatcher/RouteCreationException.cs]: What type of exception should be thrown when a constraint cannot be created?

**Comment by jbagga** [comment]
@rynowak Updated

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**